stdlib/src/binascii.rs

```diff
@@ -171,25 +179,125 @@ mod decl {
             -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1, -1,-1,-1,-1,
         ];
 
+        const PAD: u8 = 61u8;
+
+        fn decode<T: AsRef<[u8]>>(
+            input: T,
+            strict_mode: bool,
+        ) -> Result<Vec<u8>, base64::DecodeError> {
+            base64::decode_config(
+                input,
+                base64::STANDARD.decode_allow_trailing_bits(!strict_mode),
+            )
+        }
+
+        // strict mode has 3 conditions
+        // 1. can't start with pad
+        // 2. can't have discontinuous padding
+        // 3. can't end with anything other than a pad
+        // 4. must have base64 characters
+
+        fn is_valid(
+            byte_list: &Vec<u8>,
+            strict_mode: bool,
+        ) -> Result<Vec<u8>, base64::DecodeError> {
+            if strict_mode {
+                // Strict mode doesn't allow leading padding
+                if byte_list.len() > 0 && byte_list[0] == PAD {
+                    println!("false because of starting with pad");
+                    return Err(base64::DecodeError::InvalidByte(0, 61));
+                }
+
+                for (i, &el) in byte_list.iter().enumerate() {
+                    // Strict mode enforces legal characters
+                    if BASE64_TABLE[i as usize] == -1 {
+                        println!("Invalid character detected (?): {}, {:?}", i, el);
+                        return Err(base64::DecodeError::InvalidByte(i, el));
+                    }
+
+                    // Don't allow discontinuous padding - this will also assert that the last byte(s)
+                    // are pads.
+                    if el == PAD {
+                        println!("false because of discontinuous padding/not ending with a pad");
+                        let remaining_bytes: HashSet<&u8> = HashSet::from_iter(&byte_list[i..]);
+
+                        if !(remaining_bytes.is_empty()
+                            || (remaining_bytes.len() == 1 && remaining_bytes.contains(&PAD)))
+                        {
+                            return Err(base64::DecodeError::InvalidLastSymbol(
+                                byte_list.len() - 1,
+                                *byte_list.last().unwrap(),
+                            ));
+                        }
+                    }
+                }
+            }
+
+            if byte_list.len() % 4 != 0 {
+                return Err(base64::DecodeError::InvalidLength);
+            }
+
+            println!("Reached final true");
+            Ok(vec![])
+        }
+
+        let A2BBase64Args { s, strict_mode } = args;
+
         s.with_ref(|b| {
-            let decoded = if b.len() % 4 == 0 {
-                decode(b)
+            println!(
+                "{} {:?}: {:?}",
+                strict_mode,
+                b,
+                is_valid(&b.to_vec(), strict_mode)
+            );
+
+            let bytes_are_valid = is_valid(&b.to_vec(), strict_mode);
```

Comment:
A small amount of quick feedback here, this will iterate over every char and then we `iter()` through it again. You can break it apart and `for`-loop through the bytes and move everything from `is_valid` in the for loop body.

Reply:
Oh hm. Are you to saying break apart the `b` vector and loop through that? I was thinking about doing that, but didn't really understand the point of filtering out the `-1` bytes only after we've decided that `decoded` was an error.

Reply:
Yeah totally! It may take me a few more days before I can work on it, work has been crazy. Appreciate your patience.

Reply:
@DimitrisJim at the risk of reinventing the wheel, could we just filter out all `-1` bits (similar to how we do now) in `is_valid` if strict_mode is true, then pass that filtered vector to `base64::decode`? Happy to go byte by byte, was just looking at `base64::decode`'s implementation last night and there are a lot of optimizations for speed there

Reply:
Totally, whatever you're most comfortable with at the end of the day. I mentioned the CPython implementation since its usually easier working against a reference implementation.

Reply:
Ended up copying the CPython implementation. Thanks for the tip!